quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/builditems/java/QuarkusBuildItemASTVisitor.java

```diff
@@ -0,0 +1,101 @@
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+* 
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.microprofile.jdt.internal.quarkus.builditems.java;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.dom.ITypeBinding;
+import org.eclipse.jdt.core.dom.Modifier;
+import org.eclipse.jdt.core.dom.TypeDeclaration;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.eclipse.lsp4mp.commons.DocumentFormat;
+import org.eclipse.lsp4mp.jdt.core.java.diagnostics.JavaDiagnosticsContext;
+import org.eclipse.lsp4mp.jdt.core.java.validators.JavaASTValidator;
+import org.eclipse.lsp4mp.jdt.core.utils.JDTTypeUtils;
+
+import com.redhat.microprofile.jdt.internal.quarkus.QuarkusConstants;
+import com.redhat.microprofile.jdt.internal.quarkus.builditems.QuarkusBuildItemErrorCode;
+
+/**
+ * Validates <code>io.quarkus.builder.item.BuildItem</code> subclasses.
+ * <ul>
+ * <li>checks if the BuildItem is final or abstract</li>
+ * </ul>
+ */
+public class QuarkusBuildItemASTVisitor extends JavaASTValidator {
+
+	private static final String INVALID_MODIFIER = "BuildItem class %2$s%1$s%2$s must either be declared final or abstract";
+
+	@Override
+	public boolean isAdaptedForDiagnostics(JavaDiagnosticsContext context, IProgressMonitor monitor)
+			throws CoreException {
+		// Collection of diagnostics for Quarkus Build Items is done only if
+		// io.quarkus.builder.item.BuildItem is on the classpath
+		IJavaProject javaProject = context.getJavaProject();
+		return JDTTypeUtils.findType(javaProject, QuarkusConstants.QUARKUS_BUILD_ITEM_CLASS_NAME) != null;
+	}
+
+	@Override
+	public boolean visit(TypeDeclaration node) {
+		if (isBuildItem(node)) {
+			validateBuildItem(node, super.getContext());
+		}
+		return super.visit(node);
+	}
+
+	/**
+	 * Returns true if the given type declaration extends
+	 * 'io.quarkus.builder.item.BuildItem' class and false otherwise.
+	 * 
+	 * @param typeDeclaration the type declaration.
+	 * @return true if the given type declaration extends
+	 *         'io.quarkus.builder.item.BuildItem' class and false otherwise.
+	 */
+	private static boolean isBuildItem(TypeDeclaration typeDeclaration) {
+		ITypeBinding typeBinding = typeDeclaration.resolveBinding();
+		if (typeBinding != null) {
+			return extendsClass(typeBinding, QuarkusConstants.QUARKUS_BUILD_ITEM_CLASS_NAME);
+		}
+		return false;
+	}
+
+	private void validateBuildItem(TypeDeclaration typeDeclaration, JavaDiagnosticsContext context) {
+		if (isValidBuildItem(typeDeclaration)) {
+			return;
+		}
+		super.addDiagnostic(createDiagnosticMessage(typeDeclaration, context.getDocumentFormat()),
+				QuarkusConstants.QUARKUS_DIAGNOSTIC_SOURCE, typeDeclaration,

```

Comment:
This highlights the entire class body; I think it makes more sense to highlight just the name of the type.

Reply:
Indeed, I fixed that:

![image](https://github.com/user-attachments/assets/9f43ee1e-e49b-4506-ba37-417fcdc6a40b)